Rotate key service (#1918)
neicnordic/sensitive-data-archive: sda/cmd/rotatekey/rotatekey.md

# rotatekey Service

> Rotates the crypt4gh encryption key of ingested files.

Reviewer: ```suggestion
Rotates the crypt4gh encryption key of ingested file headers.
```
We are not reencrypting the entire file.

## Service Description

The rotatekey service re-encrypts the header of a file with the configured target key, and updates the database with the new header and encryption key hash.

> When running, rotatekey reads messages from the `mapping_stream` RabbitMQ queue.

Reviewer 2: is it the `mapping_stream` or the `rotatekey_stream` https://github.com/neicnordic/sensitive-data-archive/pull/1918/files#diff-19c8132437a099a64900a4e4e0f5a2f7aef5a8b7c6d43b027025d8c29c2c1698R25

> For each message, these steps are taken (errors halts progress, the message is Nack'ed, an info-error message is sent and the service moves on to the next message):

Reviewer 3: Not very clear to me. Do the steps mean those included in the parenthesis or listed in the bullet points below?

1. The message is validated as valid JSON that matches the "dataset-mapping" schema.
2. A database look-up is performed for the configured target public key hash. If the look-up fails or the key has been deprecated, an error is raised.
3. If the massage contains more than one accession IDs, an error is raised and the message is discarded as the service works on a one file per message basis.
4. The file ID is fetched from the database.
5. The key hash of the c4gh key with which the file is currently encrypted is fetched from the database and compared with the configured target key.
6. If these key hashes differ, the reencrypt service is called to re-encrypt the file header with the target key.

> 7. The file header entry in the database is updated with the new one.

Reviewer 2: Would we not want to capture that a key for a file has been rotated in the file_event_log?

Reviewer 2: Or would that be inferred by the re verification by the verify service?

Author: Good point, I should have probably mentioned in the PR description that  the field `last_modified_by` in the files table will change to `rotatekey` after  the key for a specific file has been rotated.

As the stack is at the moment we have some events driven by file status in the file_event_log so having the last status as reencrypted or similar might require some more thought. It does make sense to me though to capture this at the file_event_log so maybe we should have a discussion about this table at some point.

8. The key hash entry for the database is updated with the new one (target key).

> 9. A re-verify message is compiled, validated and sent to the archived queue so that it is consumed by verify service.

Reviewer 3: ```suggestion
9. A re-verify message is compiled, validated and sent to the archived queue so that it is consumed by the `verify` service.
```

- `BROKER_HOST`: hostname of the rabbitmq server
- `BROKER_PORT`: rabbitmq broker port (commonly `5671` with TLS and `5672` without)
- `BROKER_QUEUE`: message queue or stream to read messages from (commonly `rotatekey_stream`)
- `BROKER_USER`: username to connect to rabbitmq
- `BROKER_PASSWORD`: password to connect to rabbitmq
- `BROKER_ROUTINGKEY`: routing from a rabbitmq exchange to the rotatekey queue

### PostgreSQL Database settings

- `DB_HOST`: hostname for the postgresql database
- `DB_PORT`: database port (commonly 5432)
- `DB_USER`: username for the database
- `DB_PASSWORD`: password for the database
- `DB_DATABASE`: database name
- `DB_SSLMODE`: The TLS encryption policy to use for database connections. Valid options are:
  - `disable`
  - `allow`
  - `prefer`
  - `require`
  - `verify-ca`
  - `verify-full`

  More information is available
  [in the postgresql documentation](https://www.postgresql.org/docs/current/libpq-ssl.html#LIBPQ-SSL-PROTECTION)

  Note that if `DB_SSLMODE` is set to anything but `disable`, then `DB_CACERT` needs to be set,
  and if set to `verify-full`, then `DB_CLIENTCERT`, and `DB_CLIENTKEY` must also be set.

> - `DB_CLIENTKEY`: key-file for the database client certificate

Reviewer 3: ```suggestion
- `DB_CLIENTKEY`: key file for the database client certificate
```
`key-file` looks like a special term or variable used in the system
